### PR TITLE
Add VolumeConservationLoss to physicsnemo.experimental.losses

### LINKED ISSUE #1806: Add a generic volume-conservation loss to physicsnemo.experimental.losses
### Motivation

Matching the spatially-integrated quantity of a field over time is a useful
weak-conservation regularizer for neural-operator surrogates. PhysicsNeMo has
mesh integration utilities but no grid-tensor conservation **loss**.

### Proposed change

Add `VolumeConservationLoss` to `physicsnemo.experimental.losses`: a generic,
dimension-agnostic loss that compares the (optionally volume-weighted) spatial
integral of `pred` vs `target` at each step of a trailing feature/time axis,
using a dependency-injected data-fitting metric callable and optional
inactive-cell masking.

Named "volume conservation" (not "mass conservation") because it conserves a
volume-weighted spatial integral; no material density/porosity is applied.
Includes unit tests and a CHANGELOG entry.

### Scope

New experimental API. Part of upstreaming reusable components from the Neural
Operator Factory example. Depends on the generic data-fitting metrics.

### PR DESCRIPTION
**Pull request by wdyab**
## Summary

Adds `VolumeConservationLoss` to `physicsnemo.experimental.losses`: a generic,
dimension-agnostic weak-conservation loss. It compares the (optionally
volume-weighted) spatial integral of `pred` and `target` at each step of a
trailing feature/time axis, using a dependency-injected data-fitting `metric`
callable and an optional inactive-cell mask.

The default `metric` is `mse` (always available); any `physicsnemo.metrics.general`
function — or a relative-error metric such as the one proposed in #1746 once it
lands — can be dependency-injected. Inputs are validated (pred/target shape,
`cell_volumes` shape, and static/per-sample `mask` shape) under a
`torch.compiler.is_compiling()` guard; the masking path uses static-shape
weighting (no boolean indexing), so it is `torch.compile`-compatible.

Named **volume conservation** (not "mass conservation") on purpose: it conserves
a volume-weighted spatial integral, not true mass — no material density or
porosity is applied.

> **Stacked PR.** Depends on #1808, which creates the `physicsnemo.experimental.losses`
> package. This branch is based on `nof/pr3b-spatial-derivative-loss`, so the diff
> also shows #1808's files until it merges; the change owned by *this* PR is the top
> commit (`physicsnemo/experimental/losses/volume_conservation.py` + its test, plus
> the `__init__` export). Once #1808 merges I'll rebase this onto `main`.

Closes #1806

## Test plan

- [x] `test/experimental/losses/test_volume_conservation.py`: analytic 2D/3D
      reservoir integrals (known totals), dependency-injected metric sweep,
      volume-weighted vs uniform, multi-timestep, masking (static + per-sample),
      gradient flow, scale invariance, uniform-default, and error paths
      (pred/target, cell_volumes, and mask shape mismatches).
- [x] `interrogate` 100%, `ruff` check+format clean, doctest, license headers, CHANGELOG.

### CODE REVIEW

File: physicsnemo/experimental/losses/volume_conservation.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+            vol = torch.ones(spatial_shape, device=pred.device, dtype=pred.dtype)
+        else:
+            vol = cell_volumes
+
+        if mask is not None:
+            union_mask = self._union_mask(mask, pred)
+            vol = vol * union_mask.to(vol.dtype)
+
+        # Broadcast weights (*spatial) -> (1, *spatial, 1) over batch and time.
+        w = vol.unsqueeze(0).unsqueeze(-1)
```
[the remark this answers is not in the document]
Comment (wdyab) [reply physicsnemo/experimental/losses/volume_conservation.py]: Fixed in c34cc5d4: per-sample `(B, *spatial)` masks are now applied per sample — the weight is `vol * mask` broadcast to `(B, *spatial, 1)` rather than reduced to the batch union — so a cell inactive for a given sample no longer contributes to that sample's integral. Static `(*spatial)` masks are unchanged. Added `test_per_sample_mask_is_not_global` (a perturbation masked-out for sample 0 but active for the unperturbed sample 1 now yields 0).

File: physicsnemo/experimental/losses/derivative.py [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+                # three stencil cells (i, i+1, i+2) along ``axis`` are active.
+                mask_axis = axis - 1
+                n = union_mask.shape[mask_axis]
+                m_left = union_mask.narrow(mask_axis, 0, n - 2)
+                m_centre = union_mask.narrow(mask_axis, 1, n - 2)
+                m_right = union_mask.narrow(mask_axis, 2, n - 2)
+                deriv_mask = m_left & m_centre & m_right
+                sel = deriv_mask.unsqueeze(0).unsqueeze(-1).expand_as(d_pred)
+                p = d_pred[sel]
+                t = d_target[sel]
```
[the remark this answers is not in the document]
Comment (wdyab) [reply physicsnemo/experimental/losses/derivative.py]: Fixed in 7f1c89d1: per-sample `(B, *spatial)` masks are now applied per sample for both the pre-differencing zeroing and the stencil-safe derivative mask (no union reduction), so a stencil window touching a sample's inactive cells is excluded for that sample only. Static `(*spatial)` masks are unchanged. Added `test_per_sample_mask_is_not_global`; the boolean-index flattening keeps this correct across the batch.

**Comment by wdyab** [comment]
@peterdsharpe @mnabian @melo-gonzo for your review

**Comment by wdyab** [comment]
Closing per the offline discussion with the maintainers: these regularization losses will live in the reservoir-simulation example recipe downstream, and the generic finite-difference need will be addressed by extending physicsnemo.nn.functional.derivatives (non-periodic + batched) rather than adding a parallel implementation. Thanks for the reviews!
